Add domain-user-transfer receiver flow

lsl opened the pull request:
Fixes https://github.com/Automattic/dotcom-forge/issues/3598

## Proposed Changes

* Adds a stepper flow for receiving domain transfers

Todo:
- [x] Waiting on a backend implementation in D121090-code - can test with this currently but it's a WIP - update 2: can now use D122231-code
- [x] Contact info is validated and sent
- [x] Term agreement is checked and sent 
- [x] Term agreement details are figured out (we'll need to link to them)
- [ ] Figure out what to do post transfer, it may end up being an async process and we'll want to direct them to a checkout thank you page or something while we poll for transfer completion

## Testing Instructions

* Apply D121090-code
* Test the transfer succeeds http://calypso.localhost:3000/setup/domain-user-transfer/domain-contact-info?domain=test345678.blog

![Screenshot 2023-09-08 at 13-55-36 WordPress com](https://github.com/Automattic/wp-calypso/assets/811776/ffb77c7b-70b6-4a17-b60f-292abe8fc040)

lsl committed:
Add domain-user-transfer receiver flow

lsl committed:
Contact form detail validation

lsl committed:
Saving contact info wip

lsl committed:
Submit the contact information to the endpoint

allilevine commented on client/landing/stepper/declarative-flow/internals/steps-repository/domain-contact-info/index.tsx:
```diff
@@ -0,0 +1,142 @@
+			formattedHeader={
+				<FormattedHeader
+					id="domain-contact-info__header"
+					headerText={ __( 'Enter your contact informaiton' ) }
```
@lsl Just noting there's a typo here. 

```suggestion
					headerText={ __( 'Enter your contact information' ) }
```

lsl committed:
Update client/landing/stepper/declarative-flow/internals/steps-repository/domain-contact-info/index.tsx

Co-authored-by: Allison Levine <1689238+allilevine@users.noreply.github.com>

lsl commented:
@wensco this PR is part of solving https://github.com/Automattic/dotcom-forge/issues/3509

I was under the assumption I would need to add a checkbox to this to allow the user to "Accept" our registration agreement but looking at the checkout screenshot below it looks like I could just add a "By accepting this transfer you accept the [domain registration agreement](https://wordpress.com/automattic-domain-name-registration-agreement/) for <domain>" to the page. Is that enough for the legal aspect or is there more to it?

![Screenshot(2)](https://github.com/Automattic/wp-calypso/assets/811776/9605953e-ad23-42e3-9edd-47a78c110cfc)

There is a list of the different types of [terms we might show on the checkout](https://github.com/Automattic/wp-calypso/blob/update/domain-transfer-receive/client/my-sites/checkout/src/components/checkout-terms.tsx#L60) but I'm not sure which ones are relevant here since it's a domain transfer and not a site plan purchase.

This is what I'm thinking so far:
![Screenshot 2023-09-18 at 15-57-38 WordPress com](https://github.com/Automattic/wp-calypso/assets/811776/cbb535a2-a09c-475d-aedf-21f86f8c1d45)

If there is only one agreement they need notice about I might shift this to a single line.

lsl committed:
Replace checkbox with term info

lsl committed:
Replace checkbox with term info

lsl committed:
TransferInfo

lsl committed:
Single line term info

candy02058912 commented:
> Going to work on the design suggestions next: https://github.com/Automattic/dotcom-forge/issues/3761.

I suppose this would be covered in another PR?

lsl commented:
> I suppose this would be covered in another PR?

Yea, I think we should ship this in whatever state and iterate separately, there is a few different things to do and they're not blocking eachother.

candy02058912 reviewed (approved):
LGTM Form is validating/working as expected.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8982333

Thank you @lsl for including a screenshot in the description! This is really helpful for our translators.

wensco commented:
Hi @lsl  - thanks for the ping!

> it looks like I could just add a "By accepting this transfer you accept the [domain registration agreement](https://wordpress.com/automattic-domain-name-registration-agreement/) for " to the page. Is that enough for the legal aspect or is there more to it?

That should be sufficient but we can run things by @jennyzhu to be certain. Jenny, this question is regarding how to handle accepting the applicable domain registration agreement when User A transfers an existing domain to User B.

>There is a list of the different types of [terms we might show on the checkout](https://github.com/Automattic/wp-calypso/blob/update/domain-transfer-receive/client/my-sites/checkout/src/components/checkout-terms.tsx#L60) but I'm not sure which ones are relevant here since it's a domain transfer and not a site plan purchase.

The link 404ed for me but I think all we would need is the domain reg agreement acceptance and updated contact info (which looks to be included in the translation request so it seems those are both covered).

For the contact info, what you have in your screenshot covers most TLDs but there are some (particularly ccTLDs) that have extra fields. There's a TLD Spec that has this info that you may be able to use. 

>If there is only one agreement they need notice about I might shift this to a single line.

There can only be one agreement per individual domain name, but if a user has multiple domains and they are able to transfer more than one at a time to another user, it's possible that the different domains will have different agreements. If the flow currently only allows one user-to-user transfer at a time, then that won't be an issue.

lsl commented:
Fixed link here: https://github.com/Automattic/wp-calypso/blob/514b17be66c92110c0596f15fed3994e0393fbba/client/my-sites/checkout/src/components/checkout-terms.tsx#L60 (sorry, I was linking to the wrong branch)

Updated screen: (can access here https://wordpress.com/setup/domain-user-transfer/domain-contact-info?domain=test345678.blog)
![Screenshot(12)](https://github.com/Automattic/wp-calypso/assets/811776/5d98398b-1c2c-4806-9940-53f537a360ac)

jennyzhu commented:
Thanks @wensco for the ping and @lsl for thinking of this! Could we make it either one of these? (The first option is preferred from a legal POV).

`By clicking "Accept domain transfer," you agree to...
`
or

`By accepting the domain transfer, you agree to...`

lsl commented:
No problem, I've already merged this one in but I can fix that up in a new PR.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8982333).
